File: mobile/mobile_central/test/services/modules/products/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/products/app/use_cases.dart';
import 'package:mobile_central/services/modules/products/domain/entities.dart';
import 'package:mobile_central/services/modules/products/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// --- Manual Mock ---

class MockProductRepository implements IProductRepository {
  final List<String> calls = [];

  PaginatedResponse<Product>? getProductsResult;
  Product? getProductByIdResult;
  Product? createProductResult;
  Product? updateProductResult;
  List<ProductIntegration>? getProductIntegrationsResult;

  Exception? errorToThrow;

  GetProductsParams? capturedGetProductsParams;
  String? capturedId;
  int? capturedBusinessId;
  CreateProductDTO? capturedCreateData;
  UpdateProductDTO? capturedUpdateData;
  String? capturedDeleteId;
  String? capturedIntegrationProductId;
  AddProductIntegrationDTO? capturedAddIntegrationData;
  int? capturedRemoveIntegrationId;

  @override
  Future<PaginatedResponse<Product>> getProducts(GetProductsParams? params) async {
    calls.add('getProducts');
    capturedGetProductsParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getProductsResult!;
  }

  @override
  Future<Product> getProductById(String id, {int? businessId}) async {
    calls.add('getProductById');
    capturedId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getProductByIdResult!;
  }

  @override
  Future<Product> createProduct(CreateProductDTO data, {int? businessId}) async {
    calls.add('createProduct');
    capturedCreateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createProductResult!;
  }

  @override
  Future<Product> updateProduct(String id, UpdateProductDTO data, {int? businessId}) async {
    calls.add('updateProduct');
    capturedId = id;
    capturedUpdateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return updateProductResult!;
  }

  @override
  Future<void> deleteProduct(String id, {int? businessId}) async {
    calls.add('deleteProduct');
    capturedDeleteId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }

  @override
  Future<List<ProductIntegration>> getProductIntegrations(String productId, {int? businessId}) async {
    calls.add('getProductIntegrations');
    capturedIntegrationProductId = productId;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getProductIntegrationsResult!;
  }

  @override
  Future<void> addProductIntegration(String productId, AddProductIntegrationDTO data, {int? businessId}) async {
    calls.add('addProductIntegration');
    capturedIntegrationProductId = productId;
    capturedAddIntegrationData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }

  @override
  Future<void> removeProductIntegration(String productId, int integrationId, {int? businessId}) async {
    calls.add('removeProductIntegration');
    capturedIntegrationProductId = productId;
    capturedRemoveIntegrationId = integrationId;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }
}

// --- Helpers ---

Product _makeProduct({String id = '1', String name = 'TestProduct'}) {
  return Product(
    id: id,
    createdAt: '2026-01-01',
    updatedAt: '2026-01-01',
    businessId: 1,
    sku: 'SKU-$id',
    name: name,
    price: 10.0,
    currency: 'COP',
    stock: 5,
    manageStock: true,
    status: 'active',
    isActive: true,
  );
}

Pagination _makePagination() {
  return Pagination(
    currentPage: 1,
    perPage: 10,
    total: 1,
    lastPage: 1,
    hasNext: false,
    hasPrev: false,
  );
}

// --- Tests ---

void main() {
  late MockProductRepository mockRepo;
  late ProductUseCases useCases;

  setUp(() {
    mockRepo = MockProductRepository();
    useCases = ProductUseCases(mockRepo);
  });

  group('getProducts', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<Product>(
        data: [_makeProduct()],
        pagination: _makePagination(),
      );
      mockRepo.getProductsResult = expected;
      final params = GetProductsParams(page: 1, pageSize: 10);

      final result = await useCases.getProducts(params);

      expect(result.data.length, 1);
      expect(result.data[0].name, 'TestProduct');
      expect(mockRepo.calls, ['getProducts']);
      expect(mockRepo.capturedGetProductsParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getProductsResult = PaginatedResponse<Product>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getProducts(null);

      expect(mockRepo.capturedGetProductsParams, isNull);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Network error');

      expect(() => useCases.getProducts(null), throwsException);
    });
  });

  group('getProductById', () {
    test('delegates to repository with correct id and businessId', () async {
      mockRepo.getProductByIdResult = _makeProduct(id: '42', name: 'Found');

      final result = await useCases.getProductById('42', businessId: 5);

      expect(result.id, '42');
      expect(result.name, 'Found');
      expect(mockRepo.capturedId, '42');
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['getProductById']);
    });
  });

  group('createProduct', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateProductDTO(businessId: 1, sku: 'S', name: 'N', price: 10, stock: 5);
      mockRepo.createProductResult = _makeProduct(id: '99', name: 'N');

      final result = await useCases.createProduct(dto, businessId: 1);

      expect(result.id, '99');
      expect(mockRepo.capturedCreateData, dto);
      expect(mockRepo.capturedBusinessId, 1);
      expect(mockRepo.calls, ['createProduct']);
    });

    test('propagates repository errors', () async {
      final dto = CreateProductDTO(businessId: 1, sku: 'S', name: 'N', price: 10, stock: 5);
      mockRepo.errorToThrow = Exception('Create failed');

      expect(() => useCases.createProduct(dto), throwsException);
    });
  });

  group('updateProduct', () {
    test('delegates to repository with correct id and data', () async {
      final dto = UpdateProductDTO(name: 'Updated');
      mockRepo.updateProductResult = _makeProduct(id: '5', name: 'Updated');

      final result = await useCases.updateProduct('5', dto, businessId: 2);

      expect(result.id, '5');
      expect(result.name, 'Updated');
      expect(mockRepo.capturedId, '5');
      expect(mockRepo.capturedUpdateData, dto);
      expect(mockRepo.capturedBusinessId, 2);
      expect(mockRepo.calls, ['updateProduct']);
    });
  });

  group('deleteProduct', () {
    test('delegates to repository with correct id', () async {
      await useCases.deleteProduct('7', businessId: 3);

      expect(mockRepo.capturedDeleteId, '7');
      expect(mockRepo.capturedBusinessId, 3);
      expect(mockRepo.calls, ['deleteProduct']);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Delete failed');

      expect(() => useCases.deleteProduct('7'), throwsException);
    });
  });

  group('getProductIntegrations', () {
    test('delegates to repository with correct productId', () async {
      mockRepo.getProductIntegrationsResult = [
        ProductIntegration(
          id: 1,
          productId: '42',
          integrationId: 10,
          externalProductId: 'ext-1',
          createdAt: '2026-01-01',
          updatedAt: '2026-01-01',
        ),
      ];

      final result = await useCases.getProductIntegrations('42', businessId: 5);

      expect(result.length, 1);
      expect(result[0].productId, '42');
      expect(mockRepo.capturedIntegrationProductId, '42');
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['getProductIntegrations']);
    });
  });

  group('addProductIntegration', () {
    test('delegates to repository with correct data', () async {
      final dto = AddProductIntegrationDTO(integrationId: 10, externalProductId: 'ext-1');

      await useCases.addProductIntegration('42', dto, businessId: 5);

      expect(mockRepo.capturedIntegrationProductId, '42');
      expect(mockRepo.capturedAddIntegrationData, dto);
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['addProductIntegration']);
    });
  });

  group('removeProductIntegration', () {
    test('delegates to repository with correct ids', () async {
      await useCases.removeProductIntegration('42', 10, businessId: 5);

      expect(mockRepo.capturedIntegrationProductId, '42');
      expect(mockRepo.capturedRemoveIntegrationId, 10);
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['removeProductIntegration']);
    });
  });
}

File: front/central/src/services/modules/products/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { ProductUseCases } from './use-cases';
import { IProductRepository } from '../domain/ports';
import {
    Product,
    PaginatedResponse,
    SingleResponse,
    ActionResponse,
    CreateProductDTO,
    UpdateProductDTO,
    AddProductIntegrationDTO,
    ProductIntegrationsResponse,
    UploadImageResponse,
} from '../domain/types';

// -----------------------------------------------------------------
// Helpers: datos de prueba reutilizables
// -----------------------------------------------------------------

const makeProduct = (overrides: Partial<Product> = {}): Product => ({
    id: '1',
    created_at: '2026-03-01T00:00:00Z',
    updated_at: '2026-03-01T00:00:00Z',
    business_id: 1,
    sku: 'SKU-001',
    name: 'Producto de prueba',
    price: 50000,
    currency: 'COP',
    stock: 100,
    manage_stock: true,
    track_inventory: true,
    status: 'active',
    is_active: true,
    ...overrides,
});

const paginatedProducts: PaginatedResponse<Product> = {
    success: true,
    message: 'OK',
    data: [makeProduct()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
};

const singleProduct: SingleResponse<Product> = {
    success: true,
    message: 'OK',
    data: makeProduct(),
};

const actionSuccess: ActionResponse = { success: true, message: 'OK' };
const actionError: ActionResponse = { success: false, message: 'Error', error: 'Something went wrong' };

// -----------------------------------------------------------------
// Mock del repositorio
// -----------------------------------------------------------------

function createMockRepository(): IProductRepository {
    return {
        getProducts: vi.fn(),
        getProductById: vi.fn(),
        createProduct: vi.fn(),
        updateProduct: vi.fn(),
        deleteProduct: vi.fn(),
        uploadProductImage: vi.fn(),
        addProductIntegration: vi.fn(),
        removeProductIntegration: vi.fn(),
        getProductIntegrations: vi.fn(),
    } as unknown as IProductRepository;
}

// -----------------------------------------------------------------
// Suite principal
// -----------------------------------------------------------------

describe('ProductUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: ProductUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new ProductUseCases(repo as unknown as IProductRepository);
    });

    // ---------------------------------------------------------------
    // getProducts
    // ---------------------------------------------------------------
    describe('getProducts', () => {
        it('debería retornar la lista paginada de productos cuando el repositorio tiene éxito', async () => {
            vi.mocked(repo.getProducts).mockResolvedValue(paginatedProducts);

            const result = await useCases.getProducts({ page: 1, page_size: 10 });

            expect(result).toEqual(paginatedProducts);
            expect(repo.getProducts).toHaveBeenCalledOnce();
            expect(repo.getProducts).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            vi.mocked(repo.getProducts).mockResolvedValue(paginatedProducts);

            await useCases.getProducts();

            expect(repo.getProducts).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Fallo de base de datos');
            vi.mocked(repo.getProducts).mockRejectedValue(expectedError);

            await expect(useCases.getProducts()).rejects.toThrow('Fallo de base de datos');
        });
    });

    // ---------------------------------------------------------------
    // getProductById
    // ---------------------------------------------------------------
    describe('getProductById', () => {
        it('debería retornar un producto por su ID', async () => {
            vi.mocked(repo.getProductById).mockResolvedValue(singleProduct);

            const result = await useCases.getProductById('1');

            expect(result).toEqual(singleProduct);
            expect(repo.getProductById).toHaveBeenCalledWith('1', undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.getProductById).mockResolvedValue(singleProduct);

            await useCases.getProductById('1', 5);

            expect(repo.getProductById).toHaveBeenCalledWith('1', 5);
        });

        it('debería propagar el error cuando el producto no existe', async () => {
            vi.mocked(repo.getProductById).mockRejectedValue(new Error('Producto no encontrado'));

            await expect(useCases.getProductById('999')).rejects.toThrow('Producto no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // createProduct
    // ---------------------------------------------------------------
    describe('createProduct', () => {
        const dto: CreateProductDTO = {
            business_id: 1,
            sku: 'SKU-002',
            name: 'Nuevo Producto',
            price: 30000,
            stock: 50,
        };

        it('debería crear un producto y retornar la respuesta del repositorio', async () => {
            vi.mocked(repo.createProduct).mockResolvedValue(singleProduct);

            const result = await useCases.createProduct(dto);

            expect(result).toEqual(singleProduct);
            expect(repo.createProduct).toHaveBeenCalledOnce();
            expect(repo.createProduct).toHaveBeenCalledWith(dto, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.createProduct).mockResolvedValue(singleProduct);

            await useCases.createProduct(dto, 5);

            expect(repo.createProduct).toHaveBeenCalledWith(dto, 5);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            vi.mocked(repo.createProduct).mockRejectedValue(new Error('SKU duplicado'));

            await expect(useCases.createProduct(dto)).rejects.toThrow('SKU duplicado');
        });
    });

    // ---------------------------------------------------------------
    // updateProduct
    // ---------------------------------------------------------------
    describe('updateProduct', () => {
        const updateDto: UpdateProductDTO = { name: 'Producto Actualizado', price: 55000 };

        it('debería actualizar un producto y retornar la respuesta del repositorio', async () => {
            const updatedResponse: SingleResponse<Product> = {
                ...singleProduct,
                data: makeProduct({ name: 'Producto Actualizado', price: 55000 }),
            };
            vi.mocked(repo.updateProduct).mockResolvedValue(updatedResponse);

            const result = await useCases.updateProduct('1', updateDto);

            expect(result).toEqual(updatedResponse);
            expect(repo.updateProduct).toHaveBeenCalledWith('1', updateDto, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.updateProduct).mockResolvedValue(singleProduct);

            await useCases.updateProduct('1', updateDto, 5);

            expect(repo.updateProduct).toHaveBeenCalledWith('1', updateDto, 5);
        });

        it('debería propagar el error cuando la actualización falla', async () => {
            vi.mocked(repo.updateProduct).mockRejectedValue(new Error('Producto no encontrado'));

            await expect(useCases.updateProduct('999', updateDto)).rejects.toThrow('Producto no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // deleteProduct
    // ---------------------------------------------------------------
    describe('deleteProduct', () => {
        it('debería eliminar un producto y retornar confirmación', async () => {
            vi.mocked(repo.deleteProduct).mockResolvedValue(actionSuccess);

            const result = await useCases.deleteProduct('1');

            expect(result).toEqual(actionSuccess);
            expect(repo.deleteProduct).toHaveBeenCalledWith('1', undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.deleteProduct).mockResolvedValue(actionSuccess);

            await useCases.deleteProduct('1', 5);

            expect(repo.deleteProduct).toHaveBeenCalledWith('1', 5);
        });

        it('debería retornar respuesta de error cuando el producto no existe', async () => {
            vi.mocked(repo.deleteProduct).mockResolvedValue(actionError);

            const result = await useCases.deleteProduct('999');

            expect(result.success).toBe(false);
            expect(result.error).toBeDefined();
        });

        it('debería propagar la excepción cuando el repositorio lanza un error de red', async () => {
            vi.mocked(repo.deleteProduct).mockRejectedValue(new Error('Network error'));

            await expect(useCases.deleteProduct('1')).rejects.toThrow('Network error');
        });
    });

    // ---------------------------------------------------------------
    // uploadProductImage
    // ---------------------------------------------------------------
    describe('uploadProductImage', () => {
        it('debería subir una imagen y retornar la URL', async () => {
            const uploadResponse: UploadImageResponse = {
                success: true,
                message: 'Imagen subida',
                image_url: 'https://cdn.example.com/img.jpg',
            };
            const formData = new FormData();
            vi.mocked(repo.uploadProductImage).mockResolvedValue(uploadResponse);

            const result = await useCases.uploadProductImage('1', formData);

            expect(result).toEqual(uploadResponse);
            expect(repo.uploadProductImage).toHaveBeenCalledWith('1', formData, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            const uploadResponse: UploadImageResponse = {
                success: true,
                message: 'OK',
                image_url: 'https://cdn.example.com/img.jpg',
            };
            const formData = new FormData();
            vi.mocked(repo.uploadProductImage).mockResolvedValue(uploadResponse);

            await useCases.uploadProductImage('1', formData, 5);

            expect(repo.uploadProductImage).toHaveBeenCalledWith('1', formData, 5);
        });

        it('debería propagar el error cuando la subida falla', async () => {
            const formData = new FormData();
            vi.mocked(repo.uploadProductImage).mockRejectedValue(new Error('Archivo muy grande'));

            await expect(useCases.uploadProductImage('1', formData)).rejects.toThrow('Archivo muy grande');
        });
    });

    // ---------------------------------------------------------------
    // addProductIntegration
    // ---------------------------------------------------------------
    describe('addProductIntegration', () => {
        const integrationDto: AddProductIntegrationDTO = {
            integration_id: 10,
            external_product_id: 'ext-prod-001',
        };

        it('debería agregar una integración al producto y retornar la respuesta', async () => {
            const response: SingleResponse<any> = { success: true, message: 'OK', data: { id: 1 } };
            vi.mocked(repo.addProductIntegration).mockResolvedValue(response);

            const result = await useCases.addProductIntegration('1', integrationDto);

            expect(result).toEqual(response);
            expect(repo.addProductIntegration).toHaveBeenCalledWith('1', integrationDto, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            const response: SingleResponse<any> = { success: true, message: 'OK', data: { id: 1 } };
            vi.mocked(repo.addProductIntegration).mockResolvedValue(response);

            await useCases.addProductIntegration('1', integrationDto, 5);

            expect(repo.addProductIntegration).toHaveBeenCalledWith('1', integrationDto, 5);
        });

        it('debería propagar el error cuando falla agregar la integración', async () => {
            vi.mocked(repo.addProductIntegration).mockRejectedValue(new Error('Integración duplicada'));

            await expect(useCases.addProductIntegration('1', integrationDto)).rejects.toThrow('Integración duplicada');
        });
    });

    // ---------------------------------------------------------------
    // removeProductIntegration
    // ---------------------------------------------------------------
    describe('removeProductIntegration', () => {
        it('debería remover una integración del producto y retornar confirmación', async () => {
            vi.mocked(repo.removeProductIntegration).mockResolvedValue(actionSuccess);

            const result = await useCases.removeProductIntegration('1', 10);

            expect(result).toEqual(actionSuccess);
            expect(repo.removeProductIntegration).toHaveBeenCalledWith('1', 10, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.removeProductIntegration).mockResolvedValue(actionSuccess);

            await useCases.removeProductIntegration('1', 10, 5);

            expect(repo.removeProductIntegration).toHaveBeenCalledWith('1', 10, 5);
        });

        it('debería propagar el error cuando falla la remoción', async () => {
            vi.mocked(repo.removeProductIntegration).mockRejectedValue(new Error('Integración no encontrada'));

            await expect(useCases.removeProductIntegration('1', 99)).rejects.toThrow('Integración no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // getProductIntegrations
    // ---------------------------------------------------------------
    describe('getProductIntegrations', () => {
        it('debería retornar las integraciones de un producto', async () => {
            const integrationsResponse: ProductIntegrationsResponse = {
                success: true,
                message: 'OK',
                data: [
                    {
                        id: 1,
                        product_id: '1',
                        integration_id: 10,
                        external_product_id: 'ext-001',
                        created_at: '2026-03-01T00:00:00Z',
                        updated_at: '2026-03-01T00:00:00Z',
                    },
                ],
                total: 1,
            };
            vi.mocked(repo.getProductIntegrations).mockResolvedValue(integrationsResponse);

            const result = await useCases.getProductIntegrations('1');

            expect(result).toEqual(integrationsResponse);
            expect(repo.getProductIntegrations).toHaveBeenCalledWith('1', undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            const integrationsResponse: ProductIntegrationsResponse = {
                success: true,
                message: 'OK',
                data: [],
                total: 0,
            };
            vi.mocked(repo.getProductIntegrations).mockResolvedValue(integrationsResponse);

            await useCases.getProductIntegrations('1', 5);

            expect(repo.getProductIntegrations).toHaveBeenCalledWith('1', 5);
        });

        it('debería propagar el error cuando la consulta de integraciones falla', async () => {
            vi.mocked(repo.getProductIntegrations).mockRejectedValue(new Error('Servicio no disponible'));

            await expect(useCases.getProductIntegrations('1')).rejects.toThrow('Servicio no disponible');
        });
    });
});
